File: Python3/1091. Shortest Path in Binary Matrix.py
# 772ms, 46.14%; 14MB, 100.00%
class Solution:
    def shortestPathBinaryMatrix(self, grid: list) -> int:
        if grid[0][0] or grid[-1][-1]: return -1
        grid[-1][-1] = -1
        row, col = len(grid), len(grid[0])
        count, stack = 1, [(0, 0)]
        while stack:
            for x, y in stack:
                grid[x][y] = count
            count += 1
            tmp = set()
            for x, y in stack:
                for x1, y1 in [(i, j) for i in [x - 1, x, x + 1] for j in [y - 1, y, y + 1]]:
                    if x1 < 0 or x1 == row or y1 < 0 or y1 == col or grid[x1][y1] > 0: continue
                    tmp.add((x1, y1))
            stack = list(tmp)
        return grid[-1][-1]

File: 1001 - 1200/1091 Shortest Path in Binary Matrix.cc
// 140ms, 18.87%; 39.1MB, 100.00%
class Solution
{
private:
    const int delta[16] = { -1, -1, -1, 0, -1, 1, 0, 1, 1, 1, 1, 0, 1, -1, 0, -1 };

public:
    int shortestPathBinaryMatrix(vector<vector<int>> &grid)
    {
        if (grid[0][0] || grid.back().back()) return -1;
        int row = grid.size(), col = grid[0].size();
        grid.back().back() = -1;

        int count = 1;
        set<pair<int, int>> s;
        s.emplace(0, 0);
        while (!s.empty()) {
            for (auto &p : s) grid[p.first][p.second] = count;
            count += 1;
            set<pair<int, int>> tmp;
            for (auto &p : s) {
                for (int i = 0; i < 16; i += 2) {
                    int x = p.first + delta[i], y = p.second + delta[i + 1];
                    if (x < 0 || y < 0 || x == row || y == row || grid[x][y] > 0) continue;
                    tmp.emplace(x, y);
                }
            }
            s = move(tmp);
        }
        return grid.back().back();
    }
};
